File: Leetcode 1301-1400/1386/1386.py
# Leetcode 1386: Cinema Seat Allocation
# https://leetcode.com/problems/cinema-seat-allocation/
# Solved on 19th of August, 2026
from collections import defaultdict


class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: list[list[int]]) -> int:
        """
        Calculates the maximum number of 4-person families that can be seated together.

        Valid 4-person allocations in a 10-seat row are seats (2, 3, 4, 5), 
        (6, 7, 8, 9), or (4, 5, 6, 7).

        Parameters:
            n (int): Total number of rows in the cinema.
            reservedSeats (List[List[int]]): 2D list of [row, col] pairs representing booked seats.

        Returns:
            int: Maximum number of 4-person family groups that can be accommodated.
        """
        reservedMap = defaultdict(int)

        for row, seat in reservedSeats:
            if 2 <= seat <= 9:
                reservedMap[row] |= (1 << seat)

        totalGroups = (n - len(reservedMap)) * 2

        leftM = (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5)
        rightM = (1 << 6) | (1 << 7) | (1 << 8) | (1 << 9)
        midM = (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7)

        for rowM in reservedMap.values():
            canL = (rowM & leftM) == 0
            canR = (rowM & rightM) == 0
            
            if canL and canR:
                totalGroups += 2
            elif canL or canR or ((rowM & midM) == 0):
                totalGroups += 1

        return totalGroups

File: Leetcode 1301-1400/1386/1386.java
// Leetcode 1386: Cinema Seat Allocation
// https://leetcode.com/problems/cinema-seat-allocation/
// Solved on 19th of August, 2026
import java.util.Hashmap;
import java.util.Map;

class Solution {
    /**
     * Calculates the maximum number of 4-person families that can be seated together
     * across n rows of 10-seat aisles, accounting for reserved seats.
     *
     * Valid 4-seat contiguous blocks are: [2, 3, 4, 5], [6, 7, 8, 9], or [4, 5, 6, 7].
     *
     * @param n             The total number of rows in the cinema theater (1-indexed).
     * @param reservedSeats A 2D integer array where each element [row, col] denotes an already booked seat.
     * @return              The maximum number of 4-person family groups that can be accommodated.
     */
    public int maxNumberOfFamilies(int n, int[][] reservedSeats) {
        Map<Integer, Integer> rowReservations = new HashMap<>();
        
        for (int[] seat : reservedSeats) {
            int row = seat[0];
            int col = seat[1];
            if (col >= 2 && col <= 9) {
                rowReservations.put(row, rowReservations.getOrDefault(row, 0) | (1 << col));
            }
        }
        
        int totalGroups = (n - rowReservations.size()) * 2;
        
        for (int mask : rowReservations.values()) {
            boolean left = (mask & 60) == 0;
            boolean right = (mask & 960) == 0;
            boolean middle = (mask & 240) == 0;
            
            if (left && right) {
                totalGroups += 2;
            } else if (left || right || middle) {
                totalGroups += 1;
            }
        }
        
        return totalGroups;
    }
}